### PR TITLE
add enum for read consistency

### PR DESCRIPTION
**Pull request by generall**
Issue: https://github.com/qdrant/qdrant/issues/1372

Added both - majority and quorum with slightly different pattern

### CODE REVIEW

File: lib/collection/src/operations/consistency_params.rs [hunk 1]
```diff
@@ -0,0 +1,121 @@
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+pub struct ReadConsistencyValidationError;
+
+impl std::fmt::Display for ReadConsistencyValidationError {
+    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(formatter, "Read consistency factor cannot be less than 1")
+    }
+}
```
[the remark this answers is not in the document]
Comment (generall) [reply lib/collection/src/operations/consistency_params.rs]: renamed

[the remark this answers is not in the document]
Comment (generall) [reply lib/collection/src/operations/consistency_params.rs]: ok, no, it seems it should be

File: lib/collection/src/operations/consistency_params.rs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+                } else {
+                    Err(ValidationError)
+                }
+            }
+            ReadConsistencyShadow::Type(read_consistency_type) => {
+                Ok(ReadConsistency::Type(read_consistency_type))
+            }
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (ffuugoo) [reply lib/collection/src/operations/consistency_params.rs:80]: Another approach:

```suggestion
use crate::utils::validate::{NonZero, Validator as _};

#[derive(Debug, Deserialize, Serialize, JsonSchema, Clone, PartialEq, Eq)]
#[serde(untagged)]
pub enum ReadConsistency {
    // send N random request and return points, which present on all of them
    Factor(#[serde(deserialize_with = "NonZero::deserialize")] usize),
    Type(ReadConsistencyType),
}

// mod crate::utils::validate
pub enum NonZero {}

impl Validator<u32> for NonZero {
    type Error = &'static str;
    
    fn validate(&value: &u32) -> Result<(), Self::Error> {
        if value != 0 {
            Ok(())
        } else {
            Err("value is non-zero")
        }
    }
}

pub trait Validator<T> {
    type Error: fmt::Display;
    
    fn validate(value: &T) -> Result<(), Self::Error>;
    
    fn deserialize<'de, D>(deserializer: D) -> Result<T, D::Error>
    where
        D: serde::Deserializer<'de>,
        T: serde::Deserialize<'de>,
    {
        let value = T::deserialize(deserializer)?;
        Self::validate(&value).map_err(serde::de::Error::custom)?;
        Ok(value)
    }
}

```

**Comment by ffuugoo** [comment]
Don't like the `ReadConsistencyType` name, but can't come up with a better one, so **LGTM**.

**Review by ffuugoo — APPROVED** [review]

**Pull request merged**